src/openms/source/QC/DBSuitability.cpp

```diff
@@ -290,9 +229,320 @@ namespace OpenMS
+    { // error occured
+      OPENMS_LOG_ERROR << "An error occured while running " << adapter_name << "." << endl;
+      OPENMS_LOG_ERROR << "Standard output: " << proc_stdout << endl;
+      OPENMS_LOG_ERROR << "Standard error: " << proc_stderr << endl;
+      throw Exception::UnableToCreateFile(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "Return state was: " + static_cast<Int>(rt));
+    }
+    // search was successful
+
+    // load result
+    vector<ProteinIdentification> prot_ids;
+    vector<PeptideIdentification> pep_ids;
+    IdXMLFile comet_out;
+    comet_out.load(out_path, prot_ids, pep_ids);
+
+    // annotate target/decoy information
+    PeptideIndexing indexer;
+    FASTAContainer<TFI_Vector> proteins(fasta_data);
+    OPENMS_LOG_DEBUG << "Running PeptideIndexer functionalities ..." << endl << endl;
+    OPENMS_LOG_INFO.remove(cout); // prevent indexer from writing statistic
+    PeptideIndexing::ExitCodes indexer_exit = indexer.run(proteins, prot_ids, pep_ids);
+    OPENMS_LOG_INFO.insert(cout); // revert logging change
+    if (indexer_exit != PeptideIndexing::ExitCodes::EXECUTION_OK)
+    {
+      OPENMS_LOG_ERROR << "An error occured while trying to index the search results." << endl;
+      throw Exception::UnableToCreateFile(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "Return state was: " + static_cast<Int>(indexer_exit));
+    }
+
+    // calculate q-values
+    Param p;
+    p.setValue("use_all_hits", "true");
+    p.setValue("add_decoy_peptides", "true");
+    p.setValue("add_decoy_proteins", "true");
+
+    FalseDiscoveryRate fdr;
+    fdr.setParameters(p);
+    OPENMS_LOG_DEBUG << "Calculating q-values ..." << endl << endl;
+    fdr.apply(pep_ids);
+
+    return pep_ids;
+  }
+
+  Size DBSuitability::countIdentifications_(std::vector<PeptideIdentification> pep_ids) const
+  {
+    Size count{};
+    for (const auto& pep_id : pep_ids)
+    {
+      vector<PeptideHit> hits = pep_id.getHits();
+      if (hits.empty()) continue;
+      if (!hits[0].metaValueExists("target_decoy"))
+      {
+        throw Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No target/decoy annotation found. Make sure PeptideIndexer ran beforehand.");
+      }
+      if (hits[0].getMetaValue("target_decoy") == "decoy") continue;
+      if (!passesFDR_(hits[0],this->getParameters().getValue("FDR"))) continue;
+
+      ++count;
+    }
+    return count;
+  }
+
+  std::vector<FASTAFile::FASTAEntry> DBSuitability::getSubsampledFasta_(const std::vector<FASTAFile::FASTAEntry>& fasta_data, double ratio) const
+  {
+    if (ratio < 0 || ratio > 1)
+    {
+      throw Exception::IllegalArgument(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "Ratio has to be between 0 and 1. Aborting!");
+    }
+    Size num_AS{};
+    for (const auto& entry : fasta_data)
+    {
+      num_AS += entry.sequence.size();
+    }
+    double num_AS_written = num_AS * ratio;
+
+    random_shuffle(fasta_data.begin(), fasta_data.end());
```

Comment:
`fasta_data` is const... why does this work???

please use 
```
  std::random_device rd;
  std::mt19937 g(rd());
  std::shuffle(v.begin(), v.end(), g);
```
since `random_shuffle` is removed in C++17...